xarray/core/dataset.py

```diff
@@ -1475,14 +1475,24 @@ def __getitem__(self, key):
         else:
             return self._copy_listed(np.asarray(key))
 
+    @overload
+    def __setitem__(self, key: Iterable[Hashable], value) -> None:
```

Comment:
```suggestion
    def __setitem__(self, key: List[Hashable], value) -> None:
```
You do a `isinstance(key, list)` check below - so it has to be `List` here.